### PR TITLE
Week-3 - Server Programming (Step 6)

### PR DESCRIPTION
**Pull request by ralucatudor**
Finished step 6 (last step) which required me the following: "Requesting a Number of Comments" & "Deleting Comments" ([walkthrough](https://github.com/ralucatudor/step/blob/master/walkthroughs/week-3-server/step-6-additional-features-walkthrough.md)).

### Requesting a Number of Comments
I updated **Dataservlet** to accept a maximum number of comments in the request for doGet() and then **limit the number of comments it returns** based on the request parameter (I included `?max-comments=` request parameters is in the **query string** of the fetch URL).
To get the maximum number of comments from the user, I added an input field in the Comments section (see max-comments-form.html) and a 'Load comments' button.

Basically, **the comments aren't loaded when entering the Comments section** (as before). So now, `getComments()` only creates the comment list container in the wrapper.
Clicking the `Load Comments` button => calling `getCommentsFromServer()` which fills the container with ...max-nr... requested comments.

### Deleting Comments
I chose to add a `Delete` button alongside each comment and **delete a single comment** instead of all the comments in the database.

P.S. I also realised that the indent for script.js was 4 spaces instead of 2, so there's also that:(... I changed to 2

live URL: [rtudor-step-2020.appspot.com](http://rtudor-step-2020.appspot.com/).

### CODE REVIEW

File: portfolio/src/main/webapp/index.html [hunk 1]
```diff
@@ -3,10 +3,11 @@
   <head>
     <meta charset="UTF-8">
     <title>Raluca Tudor</title>
-    <link rel="stylesheet" href="style.css">
+    <link href="https://fonts.googleapis.com/css2?family=Parisienne&display=swap" rel="stylesheet">
     <link href="https://fonts.googleapis.com/css2?family=Press+Start+2P&display=swap" rel="stylesheet"> 
-    <script src="https://use.fontawesome.com/69431153f1.js"></script>
+    <link href="style.css" rel="stylesheet">
     <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+    <script src="https://use.fontawesome.com/69431153f1.js"></script>
```
Comment (ralucatudor) [inline portfolio/src/main/webapp/index.html:10]: Here I only did some reformatting

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -35,17 +36,32 @@
...
+    // Try to convert the value from the 'max-comments' field of the query string to int
+    try {
+      maxCommentsNumber = Integer.parseInt(request.getParameter("max-comments"));
+    } catch (NumberFormatException e) {
+      // Handle the exception
+      maxCommentsNumber = REQUESTED_COMMENTS_LIMIT;
+    }
+
+    // Assure that maxCommentsNumber is not negative
+    maxCommentsNumber = Math.max(maxCommentsNumber, 0);
```
Comment (skopecki) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: If you get a negative number I would fallback to the maximum instead. Returning 0 comments seems not very useful.

File: portfolio/src/main/webapp/max-comments-form.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<p> Choose the maximum number of comments to display below.</p>
+
+<form>
+	<label for="max-comments">Maximum number:</label>
```
Comment (skopecki) [inline portfolio/src/main/webapp/max-comments-form.html]: Indentation in this file is inconsistent.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -35,17 +36,32 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private static final int REQUESTED_COMMENTS_LIMIT = 1000;
```
Comment (skopecki) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: If you actually had 1000 comments, I think this would be excessive to display. Maybe 100 is better.

**Comment by skopecki** [comment]
For future reviews: If you have a change which is only doing formatting and touching a lot of lines of code, separate it from any other code changes. Since there are so many changed lines, it's easy to miss new code which was added or spend time reviewing code which was already there.

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -41,110 +41,159 @@ navbarList.setAttribute('class', 'navbar-list');
...
+  // Add "Comments" Heading
+  const commentsTitle = document.createElement('h1');
+  commentsTitle.textContent = 'Comments';
+  wrapper.appendChild(commentsTitle);
+
+  // Add button that redirects to the form for adding a new comment
+  const newCommentButton = document.createElement('button');
+  newCommentButton.textContent = 'Add New Comment';
+  newCommentButton.onclick = getCommentForm;
+  wrapper.appendChild(newCommentButton);
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: I think it would be better to have an html file which contains the static elements of the comments page and only fill in the comments programmatically.

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -41,110 +41,159 @@ navbarList.setAttribute('class', 'navbar-list');
...
+  const maxCommentsForm = document.createElement('div');
+  maxCommentsForm.setAttribute('id', 'max-comments-form-container');
+  // $(function() is a shorthand for jQuery(document).ready(function()
+  $(function() {
+    $('#max-comments-form-container').load('max-comments-form.html');
+  });
+  wrapper.appendChild(maxCommentsForm);
+
+  // Create container for displaying comments 
+  const commnentsContainer = document.createElement('ul');
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: Typo: commentsContainer

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -41,110 +41,159 @@ navbarList.setAttribute('class', 'navbar-list');
...
+  const newCommentButton = document.createElement('button');
+  newCommentButton.textContent = 'Add New Comment';
+  newCommentButton.onclick = getCommentForm;
+  wrapper.appendChild(newCommentButton);
+
+  // Add form where the user can pick a maximum number of comments to fetch and display
+  const maxCommentsForm = document.createElement('div');
+  maxCommentsForm.setAttribute('id', 'max-comments-form-container');
+  // $(function() is a shorthand for jQuery(document).ready(function()
+  $(function() {
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: You could use an arrow function instead: 
$(() => { ...

But not feeling strongly here.

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -41,110 +41,159 @@ navbarList.setAttribute('class', 'navbar-list');
...
-function createListElement(text) {
-    const liElement = document.createElement('li');
-    liElement.innerText = text;
-    return liElement;
+function getCommentsFromServer() {
+  // Get the maximum number of comments from the user input
+  var maxCommentsNumber = document.getElementById("max-comments-number").value;
+  if (maxCommentsNumber === null) {
+    maxCommentsNumber = 0;
+  }
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: To make this a little shorter, you could write:

var maxCommentsNumber = document.getElementById("max-comments-number").value || 0;

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -41,110 +41,159 @@ navbarList.setAttribute('class', 'navbar-list');
...
+
+  // Fetch comments from the server and add them to the DOM
+  const fetchURL = `/data?max-comments=${maxCommentsNumber}`;
+  const commnentsContainer = document.getElementById('comments-container');
+  commnentsContainer.innerHTML = '';
+  fetch(fetchURL).then(response => response.json()).then((comments) => {
```
Comment (skopecki) [inline portfolio/src/main/webapp/script.js]: For consistency, wrap response in parenthesis: ...(response) => response.json()...

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (ralucatudor) [reply portfolio/src/main/webapp/script.js]: What I mistake I've made! Won't happen again! Thank you for noticing!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Indeed. Done!

File: portfolio/src/main/webapp/script.js (hunk 8 above)
Comment (ralucatudor) [reply portfolio/src/main/webapp/script.js]: Thank you for the tip!

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (ralucatudor) [reply portfolio/src/main/webapp/script.js]: I tried this and it gave me an error in the console :(

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (ralucatudor) [reply portfolio/src/main/webapp/script.js]: Will do!

Comment (ralucatudor) [reply portfolio/src/main/webapp/script.js]: Done! What do you think? Is it better now? [Commit link](https://github.com/ralucatudor/step/pull/4/commits/0d0919116f0c379e2ae448f8328eef746bbf3ecd)

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (ralucatudor) [reply portfolio/src/main/webapp/script.js]: After adding a HTML file for the static elements of the 'Comments section', I got rid of these lines

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (skopecki) [reply portfolio/src/main/webapp/script.js]: yes, thanks :)

**Review by skopecki — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -35,17 +36,34 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private static final int REQUESTED_COMMENTS_LIMIT = 100;
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Sort comments by date from newest to oldest
+    int maxCommentsNumber;
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:43]: With this limit, how could a client get all the comments? Would it be better to:

#1 by default, return all comments.
#2 if "maxCommentsNumber" is set, only return that many comments.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -35,17 +36,34 @@
...
+
+    // Try to convert the value from the 'max-comments' field of the query string to int
+    try {
+      maxCommentsNumber = Integer.parseInt(request.getParameter("max-comments"));
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Can we check whether a parameter exist before pasing?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -35,17 +36,34 @@
...
+    } catch (NumberFormatException e) {
+      // Handle the exception
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The intent is clear here. So no need to have this comment.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -35,17 +36,34 @@
...
+      maxCommentsNumber = REQUESTED_COMMENTS_LIMIT;
+    }
+
+    // If maxCommentsNumber is negative, fallback to the maximum instead
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: If we do in the current way, we are trying to fix problems of clients. This is not a proper behavior. When the input is wrong, we need to throw an exception so that client can get an error and then fix it.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:43]: Done, just that by default, I return REQUESTED_COMMENTS_LIMIT comments (imagining I had 1 billion comments...)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 11 above)
Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 13 above)
Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done!

Comment (ralucatudor) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Also very nice point, thank you!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 14]
```diff
@@ -35,17 +36,39 @@
...
+    if (request.getParameterMap().containsKey("max-comments")) {
+      // Try to convert the value from the 'max-comments' field of the query string to int
+      try {
+        maxCommentsNumber = Integer.parseInt(request.getParameter("max-comments"));
+      } catch (NumberFormatException e) {
+        maxCommentsNumber = REQUESTED_COMMENTS_LIMIT;
+      }
+    } else {
+      response.sendError(400, "'max-comments' parameter is missing.");
+      return;
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The "max-comments" could be an optional parameter. The client should be able to choose to not set it and expect to fetch all the commetns. Or we may want to support pagination. 

When later the API has multiple parameters, we need to decide which parameters are required and which are optional. For optional ones, we need to define the default behavior ideally through API documentation.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 15]
```diff
@@ -35,17 +36,39 @@
...
+    }
+
+    // If maxCommentsNumber is negative, send error
+    if (maxCommentsNumber < 0) {
```
Comment (ChangchengXiao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:59]: What should they do when they get such an error message? It's hard for clients to link this error message with the problematic parameter. Also "max-comments" provides the maximum number of comments they want to load. The client doesn't intend to load negative number of comments. So we need to connect this error with the parameter so that clients know what's wrong. One optional idea: "max-comments cannot be negative".

**Review by ChangchengXiao — APPROVED** [review]
Overall looks good!
